File: src/main/java/com/nurkiewicz/tsclass/parser/visitors/MultiplicativeExpressionVisitor.java
package com.nurkiewicz.tsclass.parser.visitors;

import com.nurkiewicz.tsclass.antlr.parser.TypeScriptBaseVisitor;
import com.nurkiewicz.tsclass.antlr.parser.TypeScriptParser;
import com.nurkiewicz.tsclass.parser.ast.expr.Expression;
import com.nurkiewicz.tsclass.parser.ast.expr.MultiplicativeExpression;

class MultiplicativeExpressionVisitor extends TypeScriptBaseVisitor<Expression> {

    @Override
    public Expression visitMultiplicativeExpression(TypeScriptParser.MultiplicativeExpressionContext ctx) {
        final Expression right = ctx.accept(new MemberExpressionVisitor());
        if (ctx.multiplicativeExpression() != null) {
            final Expression left = ctx.multiplicativeExpression().accept(this);
            final MultiplicativeExpression.Operator op = MultiplicativeExpression.Operator.of(ctx.op.getText());
            return new MultiplicativeExpression(left, op, right);
        }
        return right;
    }

}

File: src/main/kotlin/com/nurkiewicz/tsclass/parser/visitors/MultiplicativeExpressionVisitor.kt
package com.nurkiewicz.tsclass.parser.visitors

import com.nurkiewicz.tsclass.antlr.parser.TypeScriptBaseVisitor
import com.nurkiewicz.tsclass.antlr.parser.TypeScriptParser
import com.nurkiewicz.tsclass.parser.ast.Expression
import com.nurkiewicz.tsclass.parser.ast.MultiplicativeExpression

internal class MultiplicativeExpressionVisitor : TypeScriptBaseVisitor<Expression>() {

    override fun visitMultiplicativeExpression(ctx: TypeScriptParser.MultiplicativeExpressionContext): Expression {
        val right = ctx.accept(UnaryExpressionVisitor())
        if (ctx.multiplicativeExpression() != null) {
            val left = ctx.multiplicativeExpression().accept(this)
            val op = MultiplicativeExpression.Operator.of(ctx.op.getText())
            return MultiplicativeExpression(left, op, right)
        }
        return right
    }

}
